File: ECONOROTA/api/src/routes/catalog.ts
import { Hono } from 'hono';

import { badRequest } from '../lib/errors';
import { distanceKm, isOpenNow, clock } from '../lib/geo';
import { getSettings } from '../lib/settings';
import type { AppEnv } from '../lib/types';

const catalog = new Hono<AppEnv>();
const MAX_LIMIT = 50;
const RADIUS_KM = 15;

type MarketRow = {
  id: string; name: string; address: string | null; district: string | null; city: string | null; lat: number | null; lng: number | null;
  rating: number; rating_count: number; is_open: number; image_url: string | null; delivery_fee_cents: number; min_order_cents: number;
  eta_min: number; eta_max: number; opens_at: string; closes_at: string;
};

const coords = (lat?: string, lng?: string) => {
  if (lat === undefined || lng === undefined) return null;
  const a = Number(lat);
  const b = Number(lng);
  if (!Number.isFinite(a) || !Number.isFinite(b) || Math.abs(a) > 90 || Math.abs(b) > 180) throw badRequest('Coordenadas inválidas.', 'validation');
  return { lat: a, lng: b };
};

const pageArgs = (limit?: string, offset?: string) => ({
  limit: Math.min(Math.max(Number(limit) || 20, 1), MAX_LIMIT),
  offset: Math.max(Number(offset) || 0, 0),
});

const escapeLike = (s: string) => s.replace(/[\\%_]/g, (ch) => `\\${ch}`);

catalog.get('/categories', async (c) => {
  const { results } = await c.env.DB.prepare('SELECT id, name, icon, color FROM categories ORDER BY sort, name').all();
  return c.json({ items: results }, 200, { 'Cache-Control': 'public, max-age=300' });
});

/** Mercados ativos; com lat/lng calcula distância e limita ao raio de entrega. */
catalog.get('/markets', async (c) => {
  const here = coords(c.req.query('lat'), c.req.query('lng'));
  const q = c.req.query('q')?.trim().toLowerCase().slice(0, 60);
  const onlyOpen = c.req.query('open') === '1';
  const sort = c.req.query('sort') ?? 'distance';

  const { results } = await c.env.DB.prepare(
    `SELECT id, name, address, district, city, lat, lng, rating, rating_count, is_open, image_url, delivery_fee_cents,
            min_order_cents, eta_min, eta_max, opens_at, closes_at
       FROM markets WHERE status = 'ativo'`,
  ).all<MarketRow>();

  const now = clock(c.env);
  // Entrega única do EconoRota: mesmo valor base para qualquer mercado (adicional só por mercado extra no pedido).
  const fee = (await getSettings(c.env)).delivery_base_cents;
  let items = results.map((m) => ({
    ...m,
    delivery_fee_cents: fee,
    is_open: isOpenNow(m.opens_at, m.closes_at, m.is_open, now),
    distance_km: here && m.lat != null && m.lng != null ? Math.round(distanceKm(here.lat, here.lng, m.lat, m.lng) * 10) / 10 : null,
  }));
  if (here) items = items.filter((m) => m.distance_km !== null && m.distance_km <= RADIUS_KM);
  if (q) items = items.filter((m) => m.name.toLowerCase().includes(q));
  if (onlyOpen) items = items.filter((m) => m.is_open);

  const by = {
    distance: (a: (typeof items)[0], b: (typeof items)[0]) => (a.distance_km ?? 999) - (b.distance_km ?? 999),
    rating: (a: (typeof items)[0], b: (typeof items)[0]) => b.rating - a.rating,
    fee: (a: (typeof items)[0], b: (typeof items)[0]) => a.delivery_fee_cents - b.delivery_fee_cents,
  }[sort] ?? (() => 0);
  // Abertos primeiro; depois o critério escolhido.
  items.sort((a, b) => Number(b.is_open) - Number(a.is_open) || by(a, b));
  return c.json({ items: items.slice(0, MAX_LIMIT) });
});

/** Produtos: busca, categoria, mercado, ofertas, ordenação e paginação. */
catalog.get('/products', async (c) => {
  const q = c.req.query('q')?.trim().slice(0, 80);
  const category = c.req.query('category_id');
  const market = c.req.query('market_id');
  const onSale = c.req.query('on_sale') === '1';
  const sub = c.req.query('sub')?.trim().slice(0, 40);
  const sort = c.req.query('sort') ?? (q ? 'relevance' : onSale ? 'discount' : 'price');
  const { limit, offset } = pageArgs(c.req.query('limit'), c.req.query('offset'));

  const where = ['p.is_active = 1', "m.status = 'ativo'"];
  const args: unknown[] = [];
  if (q) {
    where.push("(p.name LIKE ? ESCAPE '\\' OR p.brand LIKE ? ESCAPE '\\')");
    args.push(`%${escapeLike(q)}%`, `%${escapeLike(q)}%`);
  }
  if (category) {
    where.push('p.category_id = ?');
    args.push(category);
  }
  if (market) {
    where.push('p.market_id = ?');
    args.push(market);
  }
  if (sub) {
    where.push('p.subcategory = ?');
    args.push(sub);
  }
  if (onSale) where.push('p.promo_price_cents IS NOT NULL AND p.promo_price_cents < p.price_cents');

  const price = 'COALESCE(p.promo_price_cents, p.price_cents)';
  const discount = 'CASE WHEN p.promo_price_cents IS NULL THEN 0 ELSE (p.price_cents - p.promo_price_cents) * 1.0 / p.price_cents END';
  const order = {
    relevance: `(p.stock > 0) DESC, (p.name LIKE ? ESCAPE '\\') DESC, ${discount} DESC, ${price}`,
    price: `(p.stock > 0) DESC, ${price} ASC`,
    discount: `(p.stock > 0) DESC, ${discount} DESC, ${price} ASC`,
  }[sort];
  if (!order) throw badRequest('Ordenação inválida.', 'validation');
  const orderArgs = sort === 'relevance' ? [`${escapeLike(q ?? '')}%`] : [];

  const { results } = await c.env.DB.prepare(
    `SELECT p.id, p.market_id, m.name AS market_name, p.category_id, p.subcategory, p.name, p.brand, p.unit, p.image_url,
            p.price_cents, p.promo_price_cents, p.stock
       FROM products p JOIN markets m ON m.id = p.market_id
      WHERE ${where.join(' AND ')}
      ORDER BY ${order}
      LIMIT ? OFFSET ?`,
  )
    .bind(...args, ...orderArgs, limit + 1, offset)
    .all();

  const hasMore = results.length > limit;
  return c.json({ items: results.slice(0, limit), next_offset: hasMore ? offset + limit : null });
});

/** Subcategorias com produtos ativos numa categoria (opcionalmente só de um mercado). */
catalog.get('/categories/:id/subcategories', async (c) => {
  const market = c.req.query('market_id');
  const { results } = await c.env.DB.prepare(
    `SELECT p.subcategory AS name, COUNT(*) AS count FROM products p JOIN markets m ON m.id = p.market_id
      WHERE p.category_id = ? AND p.subcategory IS NOT NULL AND p.is_active = 1 AND m.status = 'ativo'
        ${market ? 'AND p.market_id = ?' : ''}
      GROUP BY p.subcategory ORDER BY count DESC, p.subcategory`,
  )
    .bind(...(market ? [c.req.param('id'), market] : [c.req.param('id')]))
    .all();
  return c.json({ items: results });
});

/** Sugestões enquanto digita: nomes de produtos e categorias. */
catalog.get('/search/suggest', async (c) => {
  const q = c.req.query('q')?.trim().slice(0, 60) ?? '';
  if (q.length < 2) return c.json({ suggestions: [], categories: [] });
  const like = `%${escapeLike(q)}%`;
  const [names, cats] = await c.env.DB.batch([
    c.env.DB.prepare(
      `SELECT DISTINCT lower(p.name) AS name FROM products p JOIN markets m ON m.id = p.market_id
        WHERE p.is_active = 1 AND m.status = 'ativo' AND p.name LIKE ? ESCAPE '\\' ORDER BY length(p.name) LIMIT 8`,
    ).bind(like),
    c.env.DB.prepare(
      `SELECT DISTINCT c.id, c.name, c.icon, c.color FROM categories c
         LEFT JOIN products p ON p.category_id = c.id AND p.is_active = 1
        WHERE c.name LIKE ? ESCAPE '\\' OR p.name LIKE ? ESCAPE '\\' ORDER BY c.sort LIMIT 8`,
    ).bind(like, like),
  ]);
  return c.json({
    suggestions: (names.results as { name: string }[]).map((r) => r.name),
    categories: cats.results,
  });
});

export default catalog;

File: ECONOROTA/app/lib/data/models/catalog.dart
import 'package:flutter/material.dart';

Color _hex(String? v, Color fallback) {
  if (v == null || !RegExp(r'^#[0-9a-fA-F]{6}$').hasMatch(v)) return fallback;
  return Color(int.parse('FF${v.substring(1)}', radix: 16));
}

class Category {
  const Category({required this.id, required this.name, required this.icon, this.color = const Color(0xFF6A1FC2)});

  final String id;
  final String name;
  final String icon;
  final Color color;

  factory Category.fromJson(Map<String, dynamic> j) => Category(
    id: j['id'] as String,
    name: j['name'] as String,
    icon: j['icon'] as String? ?? 'category',
    color: _hex(j['color'] as String?, const Color(0xFF6A1FC2)),
  );
}

class Market {
  const Market({
    required this.id,
    required this.name,
    required this.rating,
    this.ratingCount = 0,
    this.distanceKm,
    this.address = '',
    this.district,
    this.imageUrl,
    this.isOpen = true,
    this.deliveryFeeCents = 0,
    this.minOrderCents = 0,
    this.etaMin = 30,
    this.etaMax = 45,
    this.opensAt = '07:00',
    this.closesAt = '22:00',
    this.lat,
    this.lng,
  });

  final String id;
  final String name;
  final String address;
  final String? district;
  final String? imageUrl;
  final double? distanceKm;
  final double rating;
  final int ratingCount;
  final bool isOpen;
  final int deliveryFeeCents;
  final int minOrderCents;
  final int etaMin;
  final int etaMax;
  final String opensAt;
  final String closesAt;
  final double? lat;
  final double? lng;

  String get eta => '$etaMin–$etaMax min';

  factory Market.fromJson(Map<String, dynamic> j) => Market(
    id: j['id'] as String,
    name: j['name'] as String,
    address: j['address'] as String? ?? '',
    district: j['district'] as String?,
    imageUrl: j['image_url'] as String?,
    distanceKm: (j['distance_km'] as num?)?.toDouble(),
    rating: (j['rating'] as num?)?.toDouble() ?? 0,
    ratingCount: j['rating_count'] as int? ?? 0,
    isOpen: j['is_open'] == true || j['is_open'] == 1,
    deliveryFeeCents: j['delivery_fee_cents'] as int? ?? 0,
    minOrderCents: j['min_order_cents'] as int? ?? 0,
    etaMin: j['eta_min'] as int? ?? 30,
    etaMax: j['eta_max'] as int? ?? 45,
    opensAt: j['opens_at'] as String? ?? '07:00',
    closesAt: j['closes_at'] as String? ?? '22:00',
    lat: (j['lat'] as num?)?.toDouble(),
    lng: (j['lng'] as num?)?.toDouble(),
  );
}

class Product {
  const Product({
    required this.id,
    required this.marketId,
    required this.categoryId,
    required this.name,
    required this.unit,
    required this.priceCents,
    required this.stock,
    this.promoPriceCents,
    this.brand,
    this.imageUrl,
    this.marketName,
    this.subcategory,
    this.categoryName,
  });

  final String id;
  final String marketId;
  final String categoryId;
  final String name;
  final String? brand;
  final String unit;
  final String? imageUrl;
  final String? marketName;
  final String? subcategory;
  final String? categoryName;
  final int priceCents;
  final int? promoPriceCents;
  final int stock;

  int get finalPriceCents => promoPriceCents ?? priceCents;
  bool get inStock => stock > 0;
  bool get onSale => promoPriceCents != null && promoPriceCents! < priceCents;

  /// Desconto em % (arredondado), 0 quando não está em oferta.
  int get discountPct => onSale ? (((priceCents - promoPriceCents!) / priceCents) * 100).round() : 0;

  /// Preço por kg ou litro (ex.: "R$ 5,18/kg"), quando a unidade permite calcular.
  ({int cents, String per})? get unitPrice {
    final m = RegExp(r'^\s*([\d.,]+)\s*(kg|g|l|ml)\s*$', caseSensitive: false).firstMatch(unit);
    if (m == null) return null;
    final qty = double.tryParse(m.group(1)!.replaceAll(',', '.'));
    if (qty == null || qty <= 0) return null;
    final u = m.group(2)!.toLowerCase();
    final base = switch (u) {
      'g' => qty / 1000,
      'ml' => qty / 1000,
      _ => qty,
    };
    return (cents: (finalPriceCents / base).round(), per: u == 'g' || u == 'kg' ? 'kg' : 'L');
  }

  /// "Marca · 5 kg" ou só a unidade.
  String get detail => [if (brand?.isNotEmpty ?? false) brand, unit].join(' · ');

  factory Product.fromJson(Map<String, dynamic> j) => Product(
    id: j['id'] as String,
    marketId: j['market_id'] as String,
    categoryId: j['category_id'] as String,
    name: j['name'] as String,
    brand: j['brand'] as String?,
    unit: j['unit'] as String? ?? 'un',
    imageUrl: j['image_url'] as String?,
    marketName: j['market_name'] as String?,
    subcategory: j['subcategory'] as String?,
    categoryName: j['category_name'] as String?,
    priceCents: j['price_cents'] as int,
    promoPriceCents: j['promo_price_cents'] as int?,
    stock: j['stock'] as int? ?? 0,
  );
}

/// Página de resultados com paginação por deslocamento.
class ResultPage<T> {
  const ResultPage(this.items, [this.nextOffset]);

  final List<T> items;
  final int? nextOffset;
}
